Docs: reword the dbm.dumb introduction

erlend-aasland opened the pull request:
Also:
- consistently use correct parameter markup
- consistently use submodule name as database name
- improve accuracy of the dbm.dumb.open() spec
- remove dumbdbm class refs and replace them with generic "database object"
- use parameter list for dbm.dumb.open()


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--114550.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

erlend-aasland committed:
Docs: reword the dbm.dumb introduction

Also:

- consistently name the database "dumbdbm"
- fix broken dumbdbm class refs
- improve accuracy of the dbm.dump.open() spec

ezio-melotti commented on Doc/library/dbm.rst:
```diff
@@ -344,26 +344,28 @@ to locate the appropriate header file to simplify building this module.
+Unlike other :mod:`dbm` backends, such as :mod:`dbm.gnu` no,
+external library is required.
+As with other persistent :term:`mappings <mapping>`,
+the keys and values are always stored as :class:`bytes`.
```
```suggestion
The :mod:`dbm.dumb` module provides a persistent :class:`dict`-like
interface which is written entirely in Python.
Unlike other :mod:`dbm` backends, such as :mod:`dbm.gnu` no
external library is required.
As with other persistent :term:`mappings <mapping>`,
the keys and values are always stored as :class:`bytes`.
```
Removed an extra comma and rewrapped.

I'm not sure the link to `mappings` is useful here, since it's talking about "persistent mappings", which I assume it refers to other `dbm` modules.  Even though they are "mapping-like", the link to mapping might end up being confusing since generally mappings are not persistent nor store they keys/values as bytes.

erlend-aasland replied:
We might want to word that slightly different, yes. Removing the existing link to mappings could help. Alternatively, mention instead the other dbm backends.

erlend-aasland replied:
I'll apply the reflow suggestion first, then I'll open a new conversation for adjusting the "mappings" part.

erlend-aasland committed:
Update Doc/library/dbm.rst

Co-authored-by: Ezio Melotti <ezio.melotti@gmail.com>

erlend-aasland commented on Doc/library/dbm.rst:
```diff
@@ -344,26 +344,28 @@ to locate the appropriate header file to simplify building this module.
+interface which is written entirely in Python.
+Unlike other :mod:`dbm` backends, such as :mod:`dbm.gnu` no
+external library is required.
+As with other persistent :term:`mappings <mapping>`,
```
@ezio-melotti, how about this?

```suggestion
As with other :mod:`dbm` backends,
```

erlend-aasland replied:
Applied

erlend-aasland committed:
Merge branch 'main' into docs/dbmdumb/reword

serhiy-storchaka reviewed (commented):
I noticed it yesterday and was going to replace all "dumbdbm" with "dbm.dumb". "dumbdbm" is the name of the module in Python 2. Now it is "dbm.dumb".

serhiy-storchaka commented on Doc/library/dbm.rst:
```diff
@@ -387,7 +389,7 @@ The module defines the following:
 
    .. versionchanged:: 3.5
-      :func:`.open` always creates a new database when the flag has the value
+      :func:`open` always creates a new database when the flag has the value
```
Does it refer to built-in open() or this module's open()?

erlend-aasland replied:
It refers to `dbm.dumb.open`.

erlend-aasland committed:
Address Serhiy's initial round of review

- Don't use "dumbdbm" as a database name; instead use the submodule name

- Don't use a class name; only talk about "database objects"

- Consistently use correct markup for parameters

- Use param list for open()

erlend-aasland commented on Doc/library/dbm.rst:
```diff
@@ -346,61 +346,66 @@ to locate the appropriate header file to simplify building this module.
-   dumbdbm database is created, files with :file:`.dat` and :file:`.dir` extensions
-   are created.
+   Open a :mod:`!dbm.dumb` database.
+   The returned object behaves similar to a :term:`mapping`,
```
🤷‍♂️ 

```suggestion
   The returned database object behaves similar to a :term:`mapping`,
```

erlend-aasland commented on Doc/library/dbm.rst:
```diff
@@ -346,61 +346,66 @@ to locate the appropriate header file to simplify building this module.
 
-   The optional *flag* argument can be:
+   :param filename:
+      The basename of the database file (without extensions).
```
IMO, "without extensions" is to be understood by the term "basename". The following should suffice:

```suggestion
      The basename of the database file.
```

erlend-aasland commented on Doc/library/dbm.rst:
```diff
@@ -409,5 +414,5 @@ The module defines the following:
    .. method:: dumbdbm.close()
 
-      Close the ``dumbdbm`` database.
+      Close the "dumbdbm" database.
```
```suggestion
      Close the database.
```

erlend-aasland committed:
Update Doc/library/dbm.rst

serhiy-storchaka commented on Doc/library/dbm.rst:
```diff
@@ -346,61 +346,66 @@ to locate the appropriate header file to simplify building this module.
-The module defines the following:
+The :mod:`dbm.dumb` module provides a persistent :class:`dict`-like
+interface which is written entirely in Python.
+Unlike other :mod:`dbm` backends, such as :mod:`dbm.gnu` no
```
```suggestion
Unlike other :mod:`dbm` backends, such as :mod:`dbm.gnu`, no
```

serhiy-storchaka commented on Doc/library/dbm.rst:
```diff
@@ -346,61 +346,66 @@ to locate the appropriate header file to simplify building this module.
+interface which is written entirely in Python.
+Unlike other :mod:`dbm` backends, such as :mod:`dbm.gnu` no
+external library is required.
+As with other :mod:`dbm` backends,
+the keys and values are always stored as :class:`bytes`.
```
This note looks strange now.

In Python 2, str and unicode was interchangeable to some degree. If you use ASCII-only, it was not important that you set `db['key'] = 'value'` or `db[u'key'] = u'value'`. But it may still be useful to know that if you set an unicode value, you get back a str value. This is what about this note. It is not about the internal format, it is about the output types.

But in Python 3 bytes and str are no longer interchangeable, they are no longer different representations of the same. So this is more serious question. The database is bytes to bytes mapping by its nature, but it accepts also str as input and automatically encode them to bytes using UTF-8. This should be true for all dbm implementations, so perhaps this explanation should be move to more generic description.

erlend-aasland replied:
> This note looks strange now.

You prefer the old variant? _"As with other persistent mappings, the keys and values are always stored as bytes."_

> This should be true for all dbm implementations, so perhaps this explanation should be move to more generic description.

This is already mentioned in the generic description. It is also mentioned in the docs of every dbm submodule. If we want to address this, we should do so in a separate PR; this PR is for dbm.dumb docs only.

It is tempting to grow the scope of doc PR, and end up changing a lot of things at once. This makes for complicated docs changes where it is hard to see what changes were actually done. I prefer if we try to focus on atomic doc changes.

serhiy-storchaka replied:
I just noticed that it looks strange (both old and new variants). It could be improved in a separate PR.

erlend-aasland replied:
I created https://github.com/python/cpython/pull/114609 to follow up this.

serhiy-storchaka commented on Doc/library/dbm.rst:
```diff
@@ -346,61 +346,66 @@ to locate the appropriate header file to simplify building this module.
+      The basename of the database file (without extensions).
+      A new database creates the following files:
+      - :samp:`{filename}.dat`
+      - :samp:`{filename}.dir`
```
Would not `:file:` role be more appropriate?

erlend-aasland replied:
IMO, no. I wanted to emphasise that the parameter in question is the *basename* of the generated files. I can do that better with :samp: than with :file:

serhiy-storchaka replied:
But `{filename}.dat` _is_ a file name.

erlend-aasland replied:
Here's the render. `:samp:` to the left, `:file:` to the right. It is not rendered as `{filename}.data`, Serhiy.

![Screenshot 2024-01-26 at 15 12 28](https://github.com/python/cpython/assets/13780613/7022e765-39a8-4ea2-84ac-58357c50a134)

erlend-aasland replied:
See the Sphinx docs for more information about `:samp:`:

- https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-samp

serhiy-storchaka replied:
According to the documentation, `filename` should be rendered as a variable part.
https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-file

erlend-aasland replied:
```suggestion
      - :file:`{filename}.dat`
      - :file:`{filename}.dir`
```

serhiy-storchaka commented on Doc/library/dbm.rst:
```diff
@@ -346,61 +346,66 @@ to locate the appropriate header file to simplify building this module.
-   The optional *flag* argument can be:
+   :param filename:
+      The basename of the database file (without extensions).
+      A new database creates the following files:
```
`sync()` and `close()` *can* also create the `.bak` file.

erlend-aasland replied:
That is true. However, this is the docs for the open() method. Mentioning sync() and close() in this place does not make for a good reference text.

erlend-aasland committed:
Update Doc/library/dbm.rst

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

erlend-aasland committed:
Update Doc/library/dbm.rst

erlend-aasland commented on Doc/library/dbm.rst:
```diff
@@ -346,61 +346,66 @@ to locate the appropriate header file to simplify building this module.
-   The optional *flag* argument can be:
+   :param filename:
+      The basename of the database file (without extensions).
+      A new database creates the following files:
```
Newline required before list.

```suggestion
      A new database creates the following files:

```

erlend-aasland committed:
Update Doc/library/dbm.rst

erlend-aasland committed:
Update Doc/library/dbm.rst

serhiy-storchaka commented on Doc/library/dbm.rst:
```diff
@@ -346,61 +346,67 @@ to locate the appropriate header file to simplify building this module.
 
 
-.. function:: open(filename[, flag[, mode]])
+.. function:: open(filename, flag="c", mode=0o666)
```
Interesting, how `0o666` is rendered?

hugovk replied:
The same, as `0o666`:

![image](https://github.com/python/cpython/assets/1324225/a403a6ec-6dc4-4101-a017-b516982e67ae)


https://cpython-previews--114550.org.readthedocs.build/en/114550/library/dbm.html#module-dbm.gnu

serhiy-storchaka replied:
Great. IIRC there was a problem in the past with rendering it as 438.
